File: without-library/graph/traversal/dfs.py
import sys

# DFS without libraries
# Standard recursion is fine (built-in)

sys.setrecursionlimit(200000)

class DFS:
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.visited = [False] * (n + 1)
        self.entry = [-1] * (n + 1)
        self.exit = [-1] * (n + 1)
        self.timer = 0
    
    def dfs(self, u):
        self.visited[u] = True
        self.entry[u] = self.timer
        self.timer += 1
        
        for v in self.adj[u]:
            if not self.visited[v]:
                self.dfs(v)
                
        self.exit[u] = self.timer
        self.timer += 1

def solve():
    # Input reading example not needed for template
    pass

File: without-library/graph/traversal/dfs.cpp
#include <iostream>
#include <vector>

using namespace std;

// DFS without bits/stdc++.h

const int MAXN = 200005;
vector<int> adj[MAXN];
bool visited[MAXN];
int entry_t[MAXN], exit_t[MAXN];
int timer;

void dfs(int u) {
    visited[u] = true;
    entry_t[u] = timer++;
    
    for (size_t i = 0; i < adj[u].size(); i++) {
        int v = adj[u][i];
        if (!visited[v]) {
            dfs(v);
        }
    }
    
    exit_t[u] = timer++;
}

int main() {
    return 0;
}
